python/ray/tune/impl/tuner_internal.py

```diff
@@ -0,0 +1,178 @@
+import copy
+import os
+from typing import Any, Callable, Dict, Optional, Type, Union
+
+import ray.cloudpickle as pickle
+from ray.ml.config import RunConfig
+from ray.ml.trainer import Trainer
+from ray.tune import Experiment, TuneError
+from ray.tune.impl.utils import execute_dataset
+from ray.tune.result_grid import ResultGrid
+from ray.tune.trainable import Trainable
+from ray.tune.tune import run
+from ray.tune.tune_config import TuneConfig
+
+
+_TRAINABLE_PKL = "trainable.pkl"
+_TUNER_PKL = "tuner.pkl"
+_TRAINABLE_KEY = "_trainable"
+_PARAM_SPACE_KEY = "_param_space"
+
+
+class TunerInternal:
+    """The real implementation behind external facing ``Tuner``.
+
+    The external facing ``Tuner`` multiplexes between local Tuner and remote Tuner
+    depending on whether in Ray client mode.
+
+    In Ray client mode, external Tuner wraps ``TunerInternal`` into a remote actor,
+    which is guaranteed to be placed on head node.
+
+    TunerInternal can be constructed from fresh, in which case, `trainable` needs to
+    be provided, together with optional `param_space`, `tune_config` and `run_config`.
+
+    It can also be restored from a previous failed run (given `restore_path`).
+    """
+
+    def __init__(
+        self,
+        restore_path: str = None,
+        trainable: Optional[
+            Union[
+                str,
+                Callable,
+                Type[Trainable],
+                Type[Trainer],
+                Trainer,
+            ]
+        ] = None,
+        param_space: Optional[Dict[str, Any]] = None,
+        tune_config: Optional[TuneConfig] = None,
+        run_config: Optional[RunConfig] = None,
+    ):
+        # Restored from Tuner checkpoint.
+        if restore_path:
+            trainable_ckpt = os.path.join(restore_path, _TRAINABLE_PKL)
+            with open(trainable_ckpt, "rb") as fp:
+                trainable = pickle.load(fp)
+
+            tuner_ckpt = os.path.join(restore_path, _TUNER_PKL)
+            with open(tuner_ckpt, "rb") as fp:
+                tuner = pickle.load(fp)
+                self.__dict__.update(tuner.__dict__)
+
+            self._is_restored = True
+            self._trainable = trainable
+            self._experiment_checkpoint_dir = restore_path
+            return
+
+        # Start from fresh
+        if not trainable:
```

Comment:
why is is optional then?

Reply:
ah, if "restore_path" is given, then all the rest is not needed, including trainable, thus optional.